fe/fe-core/src/main/java/com/starrocks/sql/spm/SPMAst2SQLBuilder.java

```diff
@@ -0,0 +1,204 @@
+import com.starrocks.analysis.InPredicate;
+import com.starrocks.analysis.LiteralExpr;
+import com.starrocks.analysis.ParseNode;
+import com.starrocks.analysis.SlotRef;
+import com.starrocks.sql.analyzer.AstToSQLBuilder;
+import com.starrocks.sql.ast.JoinRelation;
+import com.starrocks.sql.ast.QueryRelation;
+import com.starrocks.sql.ast.QueryStatement;
+import com.starrocks.sql.ast.SelectList;
+import com.starrocks.sql.ast.SelectListItem;
+import com.starrocks.sql.ast.SelectRelation;
+import com.starrocks.sql.ast.TableRelation;
+import org.apache.commons.collections4.CollectionUtils;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Why not use SqlDigestBuilder?
+ * SPM need special spm function to serialize/deserialize digest
+ */
+public class SPMAst2SQLBuilder {
+    private final boolean enableHints;
+
+    private final boolean enableDigest;
+
+    private final Builder builder = new Builder();
+
+    // for SQLPlanHash
+    private final Set<Long> tables = Sets.newHashSet();
+
+    private int aggCount = 0;
+
+    private int joinCount = 0;
+
+    public SPMAst2SQLBuilder(boolean enableHints, boolean enableDigest) {
+        this.enableHints = enableHints;
+        this.enableDigest = enableDigest;
+    }
+
+    public String build(QueryRelation statement) {
+        return new QueryStatement(statement).accept(builder, null);
+    }
+
+    public String build(QueryStatement statement) {
+        return statement.accept(builder, null);
+    }
+
+    public long buildHash() {
+        // low 32 bits: 6bit(agg)-6bit(join)-6bit(tableSize)-14bit(tableId)
+        // high 32 bits: reserved
+        long hash = 0;
+        hash |= aggCount;
+        hash <<= 6;
+        hash |= joinCount;
+        hash <<= 6;
+        hash |= tables.size();
+        hash <<= 14;
+        long tableHash = tables.stream().sorted().reduce(1L, (a, b) -> 31 * a + b);
+        tableHash &= (1L << 14) - 1;
+        hash |= tableHash;
+        return hash;
+    }
+
+    // need deserialize to SQL, so extends from AstToSQLBuilder, not AstToStringBuilder/SqlDigestBuilder
+    private class Builder extends AstToSQLBuilder.AST2SQLBuilderVisitor {
+        protected Builder() {
+            super(false, false, true);
+        }
+
+        @Override
+        protected String printWithParentheses(ParseNode node) {
+            if (node instanceof SlotRef || node instanceof LiteralExpr || SPMFunctions.isSPMFunctions((Expr) node)) {
+                return visit(node);
+            } else {
+                return "(" + visit(node) + ")";
+            }
+        }
+
+        @Override
+        public String visitTable(TableRelation node, Void outerScope) {
+            if (node.getTable() != null) {
+                if (node.getTable().isNativeTable()) {
+                    tables.add(node.getTable().getId());
+                } else {
+                    tables.add((long) node.getTable().getTableIdentifier().hashCode());
```

Comment:
the `getTableIdentifier` default value of OlapTable is name . I think it best to use `getId()` for olapTable. and external table use `getTableIdentifier`.